src/main/java/edu/harvard/iq/dataverse/api/Files.java

```diff
@@ -854,7 +855,144 @@ private void exportDatasetMetadata(SettingsServiceBean settingsServiceBean, Data
+     * - locale: string specifying the locale for internationalization
+     * 
+     * The response includes:
+     * - toolUrl: the URL to access the external tool
+     * - toolName: the display name of the external tool
+     * - fileId: the ID of the file
+     * - preview: whether the URL is for preview mode
+     * 
+     * Authentication is required, and appropriate permissions are checked before generating the URL.
+     * For restricted files (including files in draft/deaccessioned datasets, embargoed files, or
+     * files with expired retention periods), the user must have DownloadFile permission.
+     * 
+     * @param crc The container request context for authentication
+     * @param fileId The ID of the file
+     * @param externalToolId The ID of the external tool
+     * @param jsonBody JSON containing optional parameters
+     * @return A Response with the external tool URL and related information
+     */
+    @POST
+    @AuthRequired
+    @Path("{id}/externalTool/{tid}/toolUrl")
+    @Consumes(MediaType.APPLICATION_JSON)
+    public Response getExternalToolUrl(@Context ContainerRequestContext crc, @PathParam("id") String fileId,
+            @PathParam("tid") long externalToolId, String jsonBody) {
+
+        boolean preview = false;
+        String locale = null;
+
+        // Parse request body for parameters
+        if (StringUtils.isNotBlank(jsonBody)) {
+            try {
+                jakarta.json.JsonObject jsonObject = JsonUtil.getJsonObject(jsonBody);
+                if (jsonObject.containsKey("preview")) {
+                    preview = jsonObject.getBoolean("preview");
+                }
+                if (jsonObject.containsKey("locale")) {
+                    locale = jsonObject.getString("locale");
+                }
+            } catch (JsonParsingException | NullPointerException e) {
+                logger.warning("Error parsing JSON: " + e.getMessage());
+                // Return a proper error response for malformed JSON
+                return error(Response.Status.BAD_REQUEST, "Invalid JSON format in request body.");
+            }
+        }
+        
+        try {
+            // Find the file
+            DataFile dataFile;
+            try {
+                dataFile = findDataFileOrDie(fileId);
+            } catch (WrappedResponse ex) {
+                return notFound("File not found for given id: " + fileId);
+            }
 
+            // Find the external tool
+            ExternalTool externalTool = externalToolService.findById(externalToolId);
+            if (externalTool == null) {
+                return error(BAD_REQUEST, "External tool not found with id: " + externalToolId);
+            }
+
+            // Check if the tool has file scope
+            if (!ExternalTool.Scope.FILE.equals(externalTool.getScope())) {
+                return error(BAD_REQUEST, "External tool does not have file scope.");
+            }
+
```

Comment:
Somewhere in here we should be calling externalToolService.meetsRequirements(). Please see this PR for details:

- https://github.com/GlobalDataverseCommunityConsortium/dataverse/pull/44